src/repr/src/relation.rs

```diff
@@ -355,6 +359,65 @@ impl proptest::arbitrary::Arbitrary for ColumnName {
     }
 }
 
+/// Stable index of a column in a [`RelationDesc`].
+#[derive(
+    Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct ColumnIndex(usize);
+
+static_assertions::assert_not_impl_all!(ColumnIndex: Arbitrary);
+
+/// The version a given column was added at.
+#[derive(
+    Clone, Copy, Debug, Eq, PartialEq, PartialOrd, Ord, Serialize, Deserialize, Hash, MzReflect,
+)]
+pub struct RelationVersion(u64);
+
+impl RelationVersion {
+    pub fn root() -> Self {
+        RelationVersion(0)
+    }
+
+    /// Returns and instance of [`ColumnVersion`] which is "one" higher than `self`.
+    pub fn bump(&self) -> Self {
+        let next_version = self
+            .0
+            .checked_add(1)
+            .expect("added more than u64::MAX columns?");
+        RelationVersion(next_version)
+    }
+
+    /// Returns the inner value.
+    ///
+    /// TODO(parkmycar): Remove this [`RelationVersion`] should be opaque.
+    pub fn inner(&self) -> &u64 {
+        &self.0
+    }
+}
+
+impl RustType<ProtoRelationVersion> for RelationVersion {
+    fn into_proto(&self) -> ProtoRelationVersion {
+        ProtoRelationVersion { value: self.0 }
+    }
+
+    fn from_proto(proto: ProtoRelationVersion) -> Result<Self, TryFromProtoError> {
+        Ok(RelationVersion(proto.value))
+    }
+}
+
+/// Metadata (other than type) for a column in a [`RelationDesc`].
+#[derive(Clone, Debug, Eq, PartialEq, Serialize, Deserialize, Hash, MzReflect)]
+struct ColumnMetadata {
+    /// Name of the column.
+    name: ColumnName,
+    /// Index into a [`RelationType`] for this column.
+    typ_idx: usize,
```

Comment:
I wanted to put `ColumnType` here instead of an index, but `RelationType` is passed around so much in the compute code that this would be a large refactor